keps/sig-network/3458-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,787 @@
+  and reports green if: that field is the zero value, or if the field is not the
+  zero value but within the timeout range. This currently handles the "general
+  API server read issue" case mentioned above, whereby if all Kube-proxies
+  encounter issues reading from the API server: the state of the healthz
+  endpoints will indicate green from all of them, given that the last event
+  received was processed successfully. It does not handle the case where one / a
+  group of Kube-proxy replicas can't read from the API server, since these
+  replicas will continue reporting green, but probably should not be. It could
+  be that Kube-proxy is using invalid credentials when communicating with the
+  API server and that the node is not actually experiencing a loss of API server
+  connectivity. This case could have pods move around without Kube-proxy being
+  able to update its rules. In the worse case: this could have Kube-proxy point
+  load balancer traffic to old pod IPs which have been recycled and are now
+  being used by new pods - leading to security concerns.
+
+C) The third case is partially covered by the validation performed Kube-proxy
+  [when it initializes](1) - which validates that the configuration for its
+  dependent networking sub-system is correctly configured. Its "runtime" checks
+  are covered by it's current health check, since, if it experiences any issues
+  updating the networking rules: it won't set `oldestPendingQueued` to the zero
+  time value and hence fail the health check.
+
+It is important to mention that all of these checks are currently being
+performed for class A) services only, and should probably be extended to class
+B) as well. The health check for class B) services should hence answer if
+endpoints are running on the node AND if Kube-proxy is healthy.
+
+[1]: https://github.com/kubernetes/kubernetes/blob/f9afd68e3b3fc05f3607ea7cb90808c3e931ba9c/pkg/proxy/apis/config/validation/validation.go#L38
+
+#### Defining what "healthy" should be for Kube-proxy
+
+This section describes how Kube-proxy might handle covering all above cases.
+
+As mentioned above, A) is not handled by Kube-proxy today.
+
+```
+<<[UNRESOLVED HC-linked-service-fields ]>>
+I don't have any good ideas on how to solve it either...running the HC server in the Kubelet instead?
+<<[/UNRESOLVED]>>
+```
+
+B) shows how Kube-proxy today plays in a good way and correctly handles the
+disaster case of: "all Kube-proxies cannot read from the API server". It also
+goes to show that changes need to be made as to handle: "some Kube-proxies
+cannot read from the API server". api-machinery currently provides informers a
+mechanism for intercepting watch errors and reacting to them, which Kube-proxy
+currently does not use. The function `SetWatchErrorHandler` allows the caller to
+define a callback which is invoked when errors are intercepted, below is an
+example of how it can be used with the `DefaultWatchErrorHandler`:
```

Comment:
Even without this we should be able to tell the difference between "failed to connect" and "connected but HTTP error" right?  Does that help?

Reply:
Might help if the watch handler only handles a subset of all HTTP status codes and we might cover more failure modes by just dealing with HTTP directly.

Reply:
punted to a later KEP